documentation/puppetserver_vs_passenger.markdown

```diff
@@ -30,6 +30,10 @@ If you have server-side Ruby code in your modules, Puppet Server will run it via
 JRuby. Generally speaking, this only affects custom parser functions and report
 processors. For the vast majority of cases, this shouldn't pose any problems, as JRuby is highly compatible with vanilla Ruby.
 
+## Ruby Compatibility For Extensions
+
+Because the Puppet Server runs master under JRuby, extension code that runs on the server -- parser functions and custom Ruby types -- needs to work under JRuby, which is a Ruby 1.9-compatible interpreter. The All-in-One Puppet 4 agent, where custom facts and types/providers run, uses MRI Ruby 2.1. So for maximum compatibility, Ruby code in your modules should run under both Ruby 1.9 and Ruby 2.1.
```

Comment:
I suggest trimming / rearranging a bit:

> Ruby extension code in your modules needs to run under both Ruby 1.9 and Ruby 2.1. This is because Puppet Server runs Puppet functions and custom resource types under JRuby (which is a Ruby 1.9-compatible interpreter), and the official `puppet-agent` releases run custom facts and types/providers under MRI Ruby 2.1.